### ISSUE TITLE
code implementation inconsistent with paper

### ISSUE THREAD

**Issue by Zerospeaker** [Dreamzz5/Simple-Align#1]
Very nice work.
However, the code implementation seems inconsistent with the paper. Without `_dequeue_and_enqueue`, everything seems ok.
Nevertheless, `_dequeue_and_enqueue` in the end of `feed_data` changes the behavior.
I understand that the data dimension should become (2B, C, H, W) after expanding them, and the reg loss uses torch.chunk(out, 2, 0) to extract image pairs from out[0:B,:,:] and out[B:2B,:,:]. 
But in `_dequeue_and_enqueue`, things are different when the queue is not full and full. When the queue is not full, it is what we want as depicted in the paper "Given features x and x′ from images with varying degradation but identical content, the regularization is defined...". When the queue is full, `torch.randperm` used in  `_dequeue_and_enqueue` makes self.lq[0:B,:,:,:] and self.lq[B:2B,:,:,:] not paired, which means "features x and x′ from images with varying degradation but identical content" is violated.
Is it a bug or not?

**Comment by Dreamzz5** [comment]
Thank you for pointing this out! After reviewing the implementation, I can confirm that this is indeed a bug. I have fixed it.

**Comment by Zerospeaker** [comment]
Thank you for your reply! I have 2 another questions:
1. The result in the paper is obtained with the queue mechanism as in RealESRGAN during training or not?
2. I am a little confused with the reg loss. 
<img width="645" alt="image" src="https://github.com/user-attachments/assets/49ccc80c-ee55-4fd8-9537-6577e6434b54">
I understand u_z represents channel intensities and it is well defined. But in `Cov(x)`, x.Tx with x'shape cxhxw means what?
In addition, `x1 = x1 - x1.mean(0, keepdims=True)
            x2 = x2 - x2.mean(0, keepdims=True)`  subtract batch mean from u_z, why is that?
<img width="634" alt="image" src="https://github.com/user-attachments/assets/3c0f403b-3e3b-450f-ad69-67f2359081d9">
It seems that random fourier features should be applied on intermediate features x, but `nolinear_out = random_fourier_features_gpu(out_).reshape(out_.shape[0], -1)` applies it on channel intensities.
Sorry to bother you, my math is not very good.

**Comment by Dreamzz5** [comment]
I'm sorry, this bug was caused by my reorganization of the code. You can obtain the pre-trained model from the readme. Regarding question two, I don't quite understand. Could you please restate it?

**Comment by Zerospeaker** [comment]
I am confused with @1 and @2
```
out = torch.nn.functional.adaptive_avg_pool2d(self.net_g.out, 1) # shape 2BxCxHxW
out = out.reshape(out.shape[0], out.shape[1], -1).permute(0, 2, 1)
x1, x2 = torch.chunk(out, 2, 0) # shape Bx1XC, represents channel intensities.
x1 = x1 - x1.mean(0, keepdims=True) # @1 subtract batch mean from x1 
x2 = x2 - x2.mean(0, keepdims=True)
l_reg, n_reg = torch.tensor(0).to(out), torch.tensor(0).to(out)
for u, v in zip(x1, x2):
    # shape of u, v : 1xC
    out_ = torch.cat([u, v])
    # @2 random_fourier_features of x1 and x2? why not of self.net_g.out as illustrated in the paper
    # the code below means getting high dimension representation of 1xC instead of x and x'. (in the paper: h_x=(u^1(x)), ..., u^n(x)) 
    nolinear_out = random_fourier_features_gpu(out_).reshape(out_.shape[0], -1) 
    a, b = torch.chunk(nolinear_out, 2, 0)
    l_reg += self.weight * reg(u, v)
    n_reg += self.weight * reg(a, b)
loss_dict['l_reg'] = l_reg
loss_dict['n_reg'] = n_reg
```

**Comment by Dreamzz5** [comment]
You are right. Adding this step mainly helps to stabilize training, as observed in experiments, and provides a slight improvement in performance.

**Comment by Zerospeaker** [comment]
OK, many many thanks. Leaving behind the codes,  in "Arbitrary style transfer in real-time with adaptive instance normalization":
<img width="716" alt="image" src="https://github.com/user-attachments/assets/3bb3b6da-ed99-4403-b2f7-1d345cf4d3f8">
u_nc and sigma_nc are well defined.
In your paper: "Then, we follow the traditions of style transfer and utilize the mean and covariance as the degradation (style) sensitive indicators [29]."
And: 
<img width="641" alt="image" src="https://github.com/user-attachments/assets/61fd86ad-478f-485d-83df-731e8bc04197">
the shape of z is CxHxW, how to compute Cov(z)?
z^T @ z is really wired for its shape  CxHxW

**Comment by Dreamzz5** [comment]
We are calculating the covariance in the channel dimension.

**Comment by Zerospeaker** [comment]
Thank you for all your reply. I'll have a try.
